trino/client.py

```diff
@@ -839,3 +768,147 @@ def decorated(*args, **kwargs):
+    """
+
+    def create(self, columns, experimental_python_types):
+        assert columns is not None
+
+        if experimental_python_types:
+            return RowMapper([self._col_func(column['typeSignature']) for column in columns])
+        return RowMapper()
+
+    def _col_func(self, column):
+        col_type = column['rawType']
+
+        if col_type == 'array':
+            return self._array_map_func(column)
+        elif col_type == 'row':
+            return self._row_map_func(column)
+        elif col_type == 'map':
+            return self._map_map_func(column)
+        elif col_type.startswith('decimal'):
+            return lambda val: Decimal(val)
+        elif col_type.startswith('double') or col_type.startswith('real'):
+            return self._double_map_func()
+        elif col_type.startswith('timestamp'):
+            return self._timestamp_map_func(column, col_type)
+        elif col_type.startswith('time'):
+            return self._time_map_func(column, col_type)
+        elif col_type == 'date':
+            return lambda val: datetime.strptime(val, '%Y-%m-%d').date()
+        else:
+            return lambda val: val
+
+    def _array_map_func(self, column):
+        elt_mapping_func = self._col_func(column['arguments'][0]['value'])
+        return lambda values: [elt_mapping_func(value) for value in values]
+
+    def _row_map_func(self, column):
+        elt_mapping_funcs = [self._col_func(arg['value']['typeSignature']) for arg in column['arguments']]
+        return lambda values: tuple(elt_mapping_funcs[idx](value) for idx, value in enumerate(values))
+
+    def _map_map_func(self, column):
+        key_mapping_func = self._col_func(column['arguments'][0]['value'])
+        value_mapping_func = self._col_func(column['arguments'][1]['value'])
+        return lambda values: {key_mapping_func(key): value_mapping_func(value) for key, value in values.items()}
+
+    def _double_map_func(self):
+        return lambda val: INF if val == 'Infinity' \
+            else NEGATIVE_INF if val == '-Infinity' \
+            else NAN if val == 'NaN' \
+            else float(val)
+
+    def _timestamp_map_func(self, column, col_type):
+        pattern = "%Y-%m-%d %H:%M:%S"
+        ms_size, ms_to_trim = self._get_number_of_digits(column)
+        if ms_size > 0:
+            pattern += ".%f"
+
+        dt_size = 20 + ms_size - ms_to_trim
+        dt_tz_offset = 20 + ms_size
+        if 'with time zone' in col_type:
+
+            if ms_to_trim > 0:
+                return lambda val: \
+                    [datetime.strptime(val[:dt_size] + val[dt_tz_offset:], pattern + ' %z')
+                        if tz.startswith('+') or tz.startswith('-')
+                        else datetime.strptime(dt[:dt_size] + dt[dt_tz_offset:], pattern)
+                                     .replace(tzinfo=pytz.timezone(tz))
+                        for dt, tz in [val.rsplit(' ', 1)]][0]
```

Comment:
while this is very concise a proper function would be more readable.

Reply:
Would it? The function relies on multiple variables computed before the lambda (e.g. `dt_tz_offset`, `tz`, etc.)